x/batch/decorator.go

```diff
@@ -0,0 +1,156 @@
+/*
+Package batch provides batch transaction support
+middleware to support multiple operations in one
+transaction
+*/
+package batch
+
+import (
+	"github.com/iov-one/weave"
+	"github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/common"
+)
+
+//----------------- Decorator ----------------
+//
+// This is just a binding from the functionality into the
+// Application stack, not much business logic here.
+
+// Decorator iterates through batch transaction messages and passes them down the stack
+type Decorator struct {
+}
+
+var _ weave.Decorator = Decorator{}
+
+// NewDecorator returns a batch transaction decorator
+func NewDecorator() Decorator {
+	return Decorator{}
+}
+
+type BatchTx struct {
+	weave.Tx
+	Msg weave.Msg
+}
+
+func (tx *BatchTx) GetMsg() (weave.Msg, error) {
+	return tx.Msg, nil
+}
+
+// Check iterates through messages in a batch transaction and passes them
+// down the stack
+func (d Decorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx,
+	next weave.Checker) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	var err error
+	msg, err := tx.GetMsg()
+	if err != nil {
+		return res, err
+	}
+
+	if batchMsg, ok := msg.(Msg); ok {
+		msgList := batchMsg.MsgList()
+		checks := make([]weave.CheckResult, len(msgList))
+		for i, msg := range msgList {
+			checks[i], err = next.Check(ctx, store, &BatchTx{Tx: tx, Msg: msg})
+			if err != nil {
+				return res, err
+			}
+		}
+		res = d.combineChecks(checks)
+		return res, err
+	}
+
+	return next.Check(ctx, store, tx)
+}
+
+// combines all data bytes as a go-amino array.
+// joins all log messages with \n
+func (*Decorator) combineChecks(checks []weave.CheckResult) weave.CheckResult {
+	datas := make([][]byte, len(checks))
+	logs := make([][]byte, len(checks))
+	var allocated, payments int64
+	for i, r := range checks {
+		datas[i] = r.Data
+		//TODO: Is this good enough conversion? Should be, in theory
+		logs[i] = []byte(r.Log)
+		allocated += r.GasAllocated
+		payments += r.GasPayment
+	}
+
+	data, _ := (&ByteArrayList{Elements: datas}).Marshal()
+	log, _ := (&ByteArrayList{Elements: logs}).Marshal()
+
+	return weave.CheckResult{
+		Data: data,
+		//TODO: Is this good enough conversion? Should be, in theory
+		Log:          string(log),
```

Comment:
We already encoded log into a protobuf as discussed with ethan, same with data.

Reply:
@ethanfrey @webmaster128 can you please chip in here, some history: https://github.com/iov-one/weave/pull/188#discussion_r234531843

Reply:
@ethanfrey @alpe @webmaster128 one thing here: I think the problem with protobuf approach is that we don't get consistent log strings, as we are using protobuf in batch, but not in other messages. Hence, we don't know what to expect.

Also, log messages are not human-readable when coded in protobuf and printed.

Reply:
Just want to add the fact, that we are not even consistent with log type in batch results. In case there is an error, we return this as a string instead.